fdbserver/masterserver.actor.cpp

```diff
@@ -1272,18 +1290,176 @@ void updateLiveCommittedVersion(Reference<MasterData> self, ReportRawCommittedVe
+			//
+			// We try to handle all of the cases listed above in a single iteration over the entries of the
+			// map (and so the logic below may not be that easy to follow).
+			bool versionCommittedAlready = false;
+			std::unordered_map<Version, Version>::iterator prevPos = self->queuedCommittedVersions.end();
+			std::unordered_map<Version, Version>::iterator nextPos = self->queuedCommittedVersions.end();
+			auto iter = self->queuedCommittedVersions.begin();
+			while (iter != self->queuedCommittedVersions.end()) {
+				if (iter->first < req.prevVersion.get()) {
+					if (req.version <= iter->second) {
+						// Case A: This version has been committed already.
+						versionCommittedAlready = true;
+						break;
+					} else {
+						if (prevPos == self->queuedCommittedVersions.end()) {
+							prevPos = iter;
+						} else if (prevPos->first < iter->first) {
+							prevPos = iter;
+						}
+					}
+				} else if (iter->first == req.prevVersion.get()) {
+					// Case A: This version has been committed already.
+					versionCommittedAlready = true;
+					ASSERT(iter->second >= req.version);
+					break;
+				} else {
+					if (nextPos == self->queuedCommittedVersions.end()) {
+						nextPos = iter;
+					} else if (iter->first < nextPos->first) {
+						nextPos = iter;
+					}
+				}
+				iter++;
+			}
+
+			if (!versionCommittedAlready) {
+				bool mergedWithPrevEntry = false;
+				bool mergedWithNextEntry = false;
+				if (prevPos != self->queuedCommittedVersions.end()) {
+					if (prevPos->second == req.prevVersion.get()) {
+						// Case B.
+						prevPos->second = req.version;
+						mergedWithPrevEntry = true;
+					}
+				}
+				if (nextPos != self->queuedCommittedVersions.end()) {
+					if (nextPos->first == req.version) {
+						if (mergedWithPrevEntry) {
+							// Case D.
+							prevPos->second = nextPos->second;
+							self->queuedCommittedVersions.erase(nextPos);
+						} else {
+							// Case C.
+							Version nextPosCommitVersion = nextPos->second;
+							self->queuedCommittedVersions.erase(nextPos);
+							ASSERT(self->queuedCommittedVersions.find(req.prevVersion.get()) ==
+							       self->queuedCommittedVersions.end());
+							self->queuedCommittedVersions[req.prevVersion.get()] = nextPosCommitVersion;
+						}
+						mergedWithNextEntry = true;
+					}
+				}
+
+				if (!mergedWithPrevEntry && !mergedWithNextEntry) {
+					// Case E: Add [prevVersion, commitVersion] to the queue
+					self->queuedCommittedVersions[req.prevVersion.get()] = req.version;
+					if (self->minQueuedCommittedVersion == invalidVersion) {
+						self->minQueuedCommittedVersion = req.prevVersion.get();
+					}
+					// Capture the wait-for-prev queue size.
+					self->waitForPrevQueueSize.addMeasurement(self->queuedCommittedVersions.size());
+				}
+			}
+		}
+	}
+
+	// Set version vector staleness state.
+	self->ssVersionVector.setStaleness(!self->queuedCommittedVersions.empty());
```

Comment:
how long could the vv be in a "stale" state?

Reply:
Hard to say. I think it depends on the latencies between proxies - tlogs and proxies - sequencer. (Also, it is possible that the latencies we see in our YCSB test environment are not a representative of the latencies in a real production deployment as the YCSB test environment appears to use nodes from multiple availability zones (and so, for example, the proxy - sequencer latency could be different for different proxies) whereas a production deployment locates all nodes in a single region.)